doc/export.conf.5
.\" Process this file with
.\" groff -man -Tascii export.conf.5
.\"
.TH EXPORT.CONF 5 "APRIL 2014" RozoFS "User Manuals"
.SH NAME
export.conf \- RozoFS export daemon configuration file
.SH DESCRIPTION
.B export.conf
is managed through
.B libconfig
<http://www.hyperrealm.com/libconfig/>
and thus complies to the
.B libconfig
grammar.

.SS layout
    It is the default transform layout used by RozoFS for each volume.
    Each file data will be transform in *forward* chunks. Any *inverse* set of them
    enables to reconstruct data (read data). The *forward* generated chunks can be 
    spread over *safe* storages.

       Valid layouts are:
           - 0 (inverse = 2, forward = 3, safe = 4);
           - 1 (inverse = 4, forward = 6, safe = 8);
           - 2 (inverse = 8, forward = 12, safe = 16);


.SS volumes
   It is a list of volumes. Each volume is identified by its vid and may contain
   a list of one or more clusters (cids). Each clusters is identified by its
   unique cid number within the configuration, and may contain at least a list of 
   *safe* storages of the same capacity (sids).

       vid: (volume identifier)
               - must be an integer.
               - must be unique for each volume managed by this exportd.

       layout: (specific layout for this volume / optional)
               - Valid layouts are 0, 1 or 2.
               - Optional field (if no layout is specified for this volume, the
                 default layout is used instead).

       cid: (cluster identifier)
               - must be an integer.
               - must be unique for each cluster managed by this exportd.

       sid: (storage identifier)
               - must be an integer (between 1 and 255).
               - must be unique within a cluster.

       georep: (whether geo-replication must occur on this volume / optional)
               - must be a boolean.
               - Optional field (default value is False).

       host: (IP or hostname of the physical node that contains the storage)
               - must be a valid IPv4 address or a valid hostname.
               - must only be set in case of a volume without geo-replication.

       site: (site number where the host resides in a multi site configuration)
               - must be a strictly positive integer.
               - must only be set in case of a volume without geo-replication. 
               - Optional field. 

       site0: (IP or hostname of the physical node that contains the storage on the main site)
               - must be a valid IPv4 address or a valid hostname.
               - must only be set in case of a volume with geo-replication. 	       

       site1: (IP or hostname of the physical node that contains the storage on the secondary site)
               - must be a valid IPv4 address or a valid hostname.
               - must only be set in case of a volume with geo-replication. 	       


.B  WARNING:
   - Use a new cluster when scaling with storage of different capacity.
   - Modifying or removing sid of storage can (will) lead to DATA LOSS !!!
   - Modifying or removing cid can (will) lead to DATA LOSS !!!

.SS filters
   This section describes some IPv4 filtering rules. It is a list of filters.
   Each filter has a name that enable the exports to reference it (see 
.B exports
section).
   A filter is a list of IPv4 subnets (IP address and mask length) with a 
.B allow
   or 
.B forbid 
rule. Incoming IP addresses for a given export are checked against the
   filter referenced by the export. The longuest prefix match subnet gives the
   rule to apply to the incoming client.
   
        filter: The name of a filter
        
        rule: allow or forbid (default is forbid)
        
        subnets: list of IPv4 subnets defining the filter behaviour.        
                A subnet contains the following parameters
                ip4subnet: is an IPv4 subnet in the format <ipv4>/<len>        
                           where <IPv4> is an IPv4 address in dot notation
                           and <len> is the subnet mask length from 1 to 32.
                rule: is the rule to apply to this subnet within {allow, forbid}           

 
.SS exports
   It's the list of rozo file system exported by the exportd daemon. Each
   exported file system is identified by its eid. root is the directory path
   where the metadata for this file system are stored. 
   squota is the number of data blocks that can be stored on a file system
   before being notified. hquota is the number maximum of data blocks that can
   be stored on a file system. vid is the volume where the file data for this
   exported filesystem will be stored. bsize defines the block size to be used 
   on this export.

       eid: (export identifier)
               - must be an integer.
               - must be unique for each exported filesystem managed by this
                 exportd.

       bsize: (defines the block size used on this export)
               - Valid bsize are within "4K", "8K", "16K", "32K".
               - default bsize is "4K".

       root: (directory path where metadata for this filesystem will be stored)
               - must be an existing directory.
               - must be unique for each filesystem managed by this exportd.

       filter: (IPv4 filter to apply to this export)
               - is optionnal (every IP is allowed when ommitted).
               - must be a string defined in the filter section.
       
       squota: (soft quota)
               - squota stands for soft quota which are non blocking quotas
                 i.e write operations will succeeded but exceeded quotas will
                 be notified in /var/run/exportd/export_[eid] monitoring file.
               - when no soft quota this field may be omitted.

       hquota: (hard quota)
               - hquota stands for blocking quotas which are blocking quotas
                 i.e every attempt to write beyond hquota will fail (EDQUOT).
               - when no hard quota this field may be omitted.

       for squota and hquota:
               - quotas are express in nb. of blocks (no suffix), Kilo, Mega,
                 Giga or Tera Bytes according to suffix K, M, G or T. Only the
                  first letter of  suffix is used that is: 128G is equivalent to
                 128Giga which in turn can be 128GigaBytes etc...
               - warning: any other suffix leads to quota express in blocks.

       vid: (volume identifier used to store file data for this filesystem)
               - must be an integer.
               - should exist in volume list.
               - several filesystem can use the same volume.

.SH EXAMPLE
.PP
.nf
.ta +3i

# sample /etc/rozofs/export.conf file
layout = 0; # Default layout 
volumes =   # List of volumes
(
    {
        # First volume using default layout
        vid    = 1; # Volume identifier
        cids=       # List of clusters of the volume
        (
            {
                # First cluster of the volume
                cid = 1; # Cluster identifier
                sids =   # List of storages of the cluster
                (
                    {sid = 01; host = "storage-node-1-1";},
                    {sid = 02; host = "storage-node-1-2";},
                    {sid = 03; host = "storage-node-1-3";},
                    {sid = 04; host = "storage-node-1-4";}
                );
            },
            {
                # Second cluster of volume
                cid = 2; # Cluster identifier
                sids =   # List of storages of the cluster
                (
                    {sid = 01; host = "storage-node-2-1";},
                    {sid = 02; host = "storage-node-2-2";},
                    {sid = 03; host = "storage-node-2-3";},
                    {sid = 04; host = "storage-node-2-4";}
                );
            }
        );
    },
    {
        # 2nd volume distributed on 4 sites
        vid    = 2; # Volume identifier
	layout = 1; # Specific layout for this volume
        cids=       # List of clusters of the volume
        (
            {
                # First cluster of the volume
                cid = 3; # Cluster identifier
                sids =   # List of storages of the cluster
                (
                    {sid = 01; host = "storage-site1-node1"; site=1;},
                    {sid = 02; host = "storage-site1-node2"; site=1;},
                    {sid = 03; host = "storage-site2-node3"; site=2;},
                    {sid = 04; host = "storage-site2-node4"; site=2;},
                    {sid = 05; host = "storage-site3-node5"; site=3;},
                    {sid = 06; host = "storage-site3-node6"; site=3;},
                    {sid = 07; host = "storage-site4-node7"; site=4;},
                    {sid = 08; host = "storage-site4-node8"; site=4;},
                );
            }
        );
    },    
    {
        # 3rd volume with georeplication
        vid = 3;       # Volume identifier
        georep = True; # Volume is geo-replicated 
        layout = 1;    # Specific layout for this volume
        cids =         # List of clusters for the volume
        (
            {
                # First cluster of this volume
                cid = 4; # Cluster identifier
                sids =   # List of storages of the cluster
                (
                    {sid = 01; site0 = "storage-node-3-1-0"; site1 = "storage-node-3-1-1";},
                    {sid = 02; site0 = "storage-node-3-2-0"; site1 = "storage-node-3-2-1";},
                    {sid = 03; site0 = "storage-node-3-3-0"; site1 = "storage-node-3-3-1";},
                    {sid = 04; site0 = "storage-node-3-4-0"; site1 = "storage-node-3-4-1";},
                    {sid = 05; site0 = "storage-node-3-5-0"; site1 = "storage-node-3-5-1";},
                    {sid = 06; site0 = "storage-node-3-6-0"; site1 = "storage-node-3-6-1";},
                    {sid = 07; site0 = "storage-node-3-7-0"; site1 = "storage-node-3-7-1";},
                    {sid = 08; site0 = "storage-node-3-8-0"; site1 = "storage-node-3-8-1";}
                );
            }
        );
    }
);

filters =
(
  {
    # Only 2 subnets are allowed 
    filter = "vol1",
    subnets =
    (
      { ip4subnet="10.10.0.0/24",     rule="allow"},
      { ip4subnet="10.20.0.0/24",     rule="allow"}
    );
  }
 ,{
    # Only one subnet is allowed except one forbiden IP address
    filter = "vol3",
    subnets =
    (
      { ip4subnet="192.168.0.0/24",   rule="allow"},
      { ip4subnet="192.168.0.22/32",  rule="forbid"}
    );
  }
);

exports = ( # List of exported filesystem

    # exported file systems on volume 1
    {eid = 1; bsize="4K"; root = "/path/to/foo1"; filter="vol1", squota="128G"; hquota="256G"; vid=1;},
    {eid = 2; bsize="4K"; root = "/path/to/foo2"; filter="vol1", squota="256G"; hquota="512G"; vid=1;},
    {eid = 3; bsize="4K"; root = "/path/to/foo3"; filter="vol1", vid=1;},

    # exported file systems on volume 2
    {eid = 4; root = "/path/to/bar1"; vid=2;}
    {eid = 5; root = "/path/to/bar2"; vid=2;}
    
    # exported file systems on volume 3
    {eid = 6; root = "/path/to/geo"; filter="vol3", vid=3;}

    # ...
);

.SH FILES
.I /etc/rozofs/export.conf (/usr/local/etc/rozofs/export.conf)
.RS
The system wide configuration file.
.\".SH ENVIRONMENT
.\".SH DIAGNOSTICS
.\".SH BUGS
.SH AUTHOR
Fizians <http://www.fizians.com>
.SH "SEE ALSO"
.BR rozofs (7),
.BR exportd (8)


debian/changelog
rozofs (1.6.alpha4) stable; urgency=low

  * (rozofs) See git log for changes

 -- Fizians S.A.S. <devel@rozofs.com>  Mon, 01 Jan 2014 00:00:00 +0100


manager/rozofs/cli/export.py
# -*- coding: utf-8 -*-
#
# Copyright (c) 2013 Fizians SAS. <http://www.fizians.com>
# This file is part of Rozofs.
#
# Rozofs is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published
# by the Free Software Foundation, version 2.
#
# Rozofs is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see
# <http://www.gnu.org/licenses/>.

from rozofs.core.platform import Platform, Role
from rozofs.cli.output import ordered_puts


def create(platform, args):
    platform.create_export(args.vid[0], args.name, None, args.squota, args.hquota)

def update(platform, args):
    platform.update_export(args.eid[0], None, None, args.squota, args.hquota)

def remove(platform, args):
    if not args.eids:
        args.eids = None

    platform.remove_export(args.eids, args.force)

def get(platform, args):

    e_host = platform._active_export_host
    configurations = platform.get_configurations([e_host], Role.EXPORTD)

    configuration = configurations[e_host][Role.EXPORTD]

    # Check exception
    if isinstance(configuration, Exception):
        raise type(configuration)(str(configuration))

    eids_l = []
    if not args.eids:
        eids_l = configuration.exports.keys()
    else:
        eids_l = args.eids

    list_l = {}
    exports_l = []
    for eid in eids_l:

        if eid not in configuration.exports:
            raise Exception("Unknown export with eid=%d." % eid)

        econfig = configuration.exports[eid]
        export_l = []
        export_l.append({'vid': econfig.vid})
        export_l.append({'root': econfig.root})
        if econfig.md5 != '':
            export_l.append({'md5': econfig.md5})
        export_l.append({'squota': econfig.squota})
        export_l.append({'hquota': econfig.hquota})

        exports_l.append({'EXPORT ' + str(eid): export_l})

    list_l.update({'EXPORTS': exports_l})
    ordered_puts(list_l)

def dispatch(args):

    p = Platform(args.exportd, Role.EXPORTD)

    globals()[args.action.replace('-', '_')](p, args)
